### PR TITLE
Retry v2 checkpoint downloads in the case of "Failure when receiving data from the peer" error message

### PR DESCRIPTION
**Pull request by nick4598**
This is a short term solution to a problem surrounding Transient Fault handling with v2 checkpoints. In the long term, we expect to work with SQLite and hopefully have them provide a way for us to better control what we do in the event of an error when downloading blocks. 

Resumable downloads are supported by blockcachevfs, meaning that we will not be repeating the same block downloads when retrying these v2 checkpoint downloads.

### CODE REVIEW

**Comment by nick4598** [comment]
Could this change also be backported to 2.19.x? I'm unsure how backporting works. Theres a recent github discussion where a connector is unable to download a v2 checkpoint due to this transient "Failure when receiving data from the peer" error. That user is on 2.19.28 at the moment.

**Comment by pmconne** [comment]
> Could this change also be backported to 2.19.x?

Are V2 checkpoints enabled in 2.x?

**Comment by calebmshafer** [comment]
> 

I think in that case we should backport this to work around the issues users are having with downloading iModels.

File: core/backend/src/CheckpointManager.ts [hunk 1]
```diff
@@ -268,16 +268,59 @@ export class V1CheckpointManager {
...
   public static readonly onDownloadV2 = new BeEvent<(job: DownloadJob) => void>();
   public static getKey(checkpoint: CheckpointProps) { return `${checkpoint.iModelId}:${checkpoint.changeset.id}`; }
 
+  /**
+   * This function expects that the "functionToRetry" throws an instance of RetryableError and just any other error otherwise if its a non retryable error. This is to allow for the case where
+   * we may not want to retry every type of error produced by the function.
+   * @param numAttempts number of times to attempt the function
+   * @param functionToRetry function to call during attempts
+   */
+  private static async withAttempts(numAttempts: number, functionToRetry: () => Promise<void>): Promise<void> {
```
Comment (kabentley) [inline core/backend/src/CheckpointManager.ts:288]: I'm worried about what this will do to the progress function. Please let me look at this more closely.

**Review by kabentley — CHANGES_REQUESTED** [review]

File: core/backend/src/CheckpointManager.ts (hunk 1 above)
Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:288]: see https://github.com/iTwin/itwinjs-core/pull/3912

**Review by kabentley — APPROVED** [review]

**Comment by nick4598** [comment]
@mergifyio backport release/2.19.x

**Pull request merged**